### ISSUE TITLE
Consistent naming of backing properties

### ISSUE THREAD

**Issue by joozek78** [Starcounter/Starcounter#3827]
With `DefaultTemplate.<property>.Bind = nameof(<backing-property>)` you can set the property that Starcounter will use to calculate the value of <property>. We seem to have no agreed upon idea on how to name those 'backing properties'. I've already seen at least three different conventions.

We should develop a common naming convention for our own convenience and to avoid confusion in apps developers'

We could just add an underscore (`_`) at the end of the name or add some special word (`Calculated` is what I've come up with, but IMO it's poor)

After we discuss it here and have a decision we should put it into guidelines and docs (where we use this pattern)

CC @warpech @tomalec @alemoi @Mnosek @miyconst @diliger

**Comment by miyconst** [comment]
Me and @ubbeK in Palma project just add `Bind` in the end.

So it looks like this:

```cs
static IndexPage() {
    DefaultTemplate.Name.Bind = nameof(NameBind);
}

public string NameBind {
    get {
        return "A value";
    }
}
```

**Comment by miyconst** [comment]
I dislike the `_` solution in front or in the end of the name, the `Calculated` is fine, but just too long for me.

**Comment by diliger** [comment]
I would follow `Bind` postfix to the name.

**Comment by chrhol** [comment]
Also look at this issue: #2964. You can get automatic binding to properties in code-behind by using the same name.

**Comment by per-samuelsson** [comment]
> Also look at this issue: #2964. You can get automatic binding to properties in code-behind by using the same name.

More specifically, [here](https://github.com/Starcounter/Starcounter/issues/2964#issuecomment-166563857) we discuss a suggested pattern.

@joozek78, does that apply to what you guys try to accomplish here? If not, can you highlight the problems you find with the pattern we discussed / proposed earlier?

**Comment by joozek78** [comment]
Woah, you're saying we don't have to use `DefaultTemplate.<property>.Bind`? That's cool. So why do even do that? I thought that was the only possible way?

**Comment by Mnosek** [comment]
> I thought that was the only possible way?

Me too!

I've just checked it on pnext and it simply works fine! That's a great enhancement. Thanks @per-samuelsson for explanation! So if it is possible we should definitely use a new feature, but for the backward compatibility if necessary - I would vote for `Bind` postfix as a pattern.

**Comment by ubbeK** [comment]
> That's cool. So why do even do that

Weren't we suppose to remove the _magic_ from the `JSON`-files? https://github.com/Starcounter/Starcounter/issues/3443

**Comment by joozek78** [comment]
True. After initial hype over easy-to-use I try to cool down to be more reasonable.

On the one hand this behavior is truly magic - VS will complain that property is already defined, but your code will build and work regardless. You have virtually no chance to discover this feature by yourself. Binding will be different in a way that is not obvious, but have potentially profound consequences (refreshing every time a patch is sent).

On the other hand having to create a separate backing property that can't be named properly (the best name is already taken in json) is inconvenient. But is it inconvenient enough to justify magic? I leaning towards no.

---

I think we agreed that whenever we use `DefaultTemplate.Name.Bind` we should add `Bind` suffix to property name. What's left is whether or not we should use magic here or explicitly configure the binding. IMO we should always use explicit binding in Samples (magic can be difficult to understand and replicate) and in the docs describe the magic as a developer's choice.

**Comment by per-samuelsson** [comment]
@ubbeK @joozek78 

Nice to see you guys reflect on this. I'm not 100% sure on the right approach here either.

I'm normally against most magic, and often vote for a little more explicit approach that better illustrate what actually goes on. But there are of course borderline cases. I mean, some things based on common-sense *convensions*. Compare for example Maven w/ MsBuild - I would vote for Mavens approach every day.

In this context, we already have a naming-comparison convention: that when you make a view model bound to a database object. Very common, and we haven't heard complaints about that it's hard to understand that.

Just curious: in what way do you see that name-name contention mapping is different between having a backend database object versus a code-behind class?

**Comment by per-samuelsson** [comment]
Continuing from previous comment, to illustrate what I mean:

**Foo.cs**
```cs
[Database] public class Foo {
  public string Bar;
}
```

**FooJson.json**
```json
{
  "Bar": "",
  "Fubar": ""
}
```

**FooJson.json.cs**
```cs
public partial class FooJson : Json, IBound<Foo> {
  public string Fubar {
    get { return "Baz"; }
  }
}
```

**Comment by miyconst** [comment]
>... VS will complain that property is already defined, but your code will build and work regardless.

This is really frustrating. If we can rid of this error, then I would say that the magic is not bigger than between JSON view-model and database objects binding, as pointed out by @per-samuelsson.

A nice way to resolve this issue would be inheritance. 

- Weaver generates an abstract class with all the properties in there. 
- The code-behind class inherits the base class. 
- Developer overrides any property.

```json
{
    "Html": "IndexPage.html",
    "Title": ""
}
```

```cs
public abstract class BaseIndexPage : Json {
    public virtual string Title { get; protected set; }
}
```

```cs
public class IndexPage : BaseIndexPage {
    public override string Title {
        get {
            return "A new value";
        }
        set {
             throw new NotImplementedException("");
        }
    }
}
```

If we can't remove the error, then I will keep using and suggesting to define bindings explicitly.

**Comment by chrhol** [comment]
> On the one hand this behavior is truly magic - VS will complain that property is already defined, but your code will build and work regardless.

> > This is really frustrating.

I don't get this. What complain is this? Is it reported it anywhere? If the property exists in code-behind, it will not exist in the generated code so there should be no warning or error.

**Comment by miyconst** [comment]
>On the one hand this behavior is truly magic - VS will complain that property is already defined, but your code will build and work regardless.

@joozek78 could you please explain what you mean by this? I just have tried it myself, and it works flawlessly without any complains from Visual Studio.

And I have to say, I really like it.

**Comment by per-samuelsson** [comment]
> I don't get this.

FWIW, me neither.

**Comment by joozek78** [comment]
At first your code is wrong - weaver already generated partial class
definition with the property. VS highlights your definition as duplicate.
Only after you run build and weaver regenerates the partial definition it
omits the property you defined. At least that was the experience for me,
I'll check again this evening or tomorrow in case I remember something wrong

**Comment by chrhol** [comment]
Thanks for the explanation @joozek78. I understand what you mean now.

Just to be clear of definitions, it's not actually the weaver (or what we call the weaver) that is responsible for this part. There is a custom MsBuild task that is responsible for generating the code for TypedJSON. 

The task is triggered only when you save the json or code-behind file, so you are correct. If you first add the property in the json-file (and save) and then add it in code-behind, VS will complain about it until you *save* the file or build the project.

We have looked into having the task being triggered using in-memory objects (for VS that is) so you get updated intellisense without saving and it's possible, since other extensions use this functionality. However, from what we found, is that it's not easy and will require quite a lot of effort. So I'm not sure it's worth it, especially now that we're moving to .Net Core and we will most likely need to find some other way to support the codegeneration and intellisense in that version.

**Comment by per-samuelsson** [comment]
>So **I'm not sure it's worth it**, especially now that we're moving to .Net Core and we will most likely need to find some other way to support the codegeneration and intellisense in that version.

If it's not very low-hanging, I would say it it's not.

**Comment by per-samuelsson** [comment]
> If you first add the property in the json-file (and save) and then add it in code-behind, VS will complain about it until you *save* the file or build the project.

Made a separate issue about that: #3829. So we can have this issue be about convention, and have a separate issue to refer to if anyone report about this problem in the future.
